### ISSUE TITLE
UseDbContext with UseProjection produces Middleware pipeline order for the field `Query.xxxxxx` is invalid

### ISSUE THREAD

**Issue by kevpye-fabdata** [ChilliCream/hotchocolate#4377]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Hi, i'm very new to using hotchocolate so maybe I am doing something totally wrong but it seems like the tool is not correctly determining the order of the attributes that are being applied to my root query class. 

This is a day 0 project so very small amounts of code. I am using entity framework, net 5 and the latest HotChocolate packages - 12.1.1. I have two simple models - continent and country. Continents can contain many countries and so I am trying to use projections to tell the framework to walk the graph.

If we ignore everything to do with EF config and even the country class the following setup breaks the schema:

Startup.cs
```
public void ConfigureServices(IServiceCollection services)
        {
            services
                .AddPooledDbContextFactory<AppDbContext>(options => {
                    options.UseSqlServer(Configuration.GetConnectionString("MYDBNAME"));
                })
                .AddGraphQLServer()                
                .AddQueryType<Query>()
                .AddProjections();
                
        }
```

Query.cs
```
public class Query
    {
        [UseDbContext(typeof(AppDbContext))]
        [UseProjection]
        public IQueryable<Continent> GetContinent([ScopedService] AppDbContext context) {
            return context.Continents;
        }
    }
```

Continent.cs
```
public class Continent {
        [Key]
        public int RecordID { get; set; }
        
        public string ContinentCode { get; set; }
        
        public string ContinentName { get; set; }

        public ICollection<Country> Countries { get; set; } = new List<Country>();
    }
```

With this configuration when I send a request to the server it tells me that the order of the attributes applied to my GetContinents function in my Query class is incorrect - the error states "Your order is: UseProjection -> UseDbContext.." when clearly in the code the UseDbContext is the first attribute. If I remove the UseProjection attribute then the schema doesnt generate an error.

I have included the full error log in the relevant error log section

### Steps to reproduce

1. Download the source code from https://github.com/kevpye-fabdata/HotChocolateErrorDemo
2. Run the demo in debug mode
3. Visit the graphql endpoint in the browser - note the the red dot in cake pop that a valid scema was not obtained
3. See the contents of the debug output showing the log output and error I am seeing

p.s. you dont need to worry about setting up the database as the application will not generate a valid schema.

### Relevant log output

```shell
HotChocolate.SchemaException: For more details look at the `Errors` property.

1. The middleware pipeline order for the field `Query.continent` is invalid. Middleware order is important especially with data pipelines. The correct order of a data pipeline is as follows: UseDbContext -> UsePaging -> UseProjection -> UseFiltering -> UseSorting. You may omit any of these middleware or have other middleware in between but you need to abide by the overall order. Your order is: UseProjection -> UseDbContext.... (HotChocolate.Types.ObjectType<ConferencePlanner.GraphQL.GraphQL.Query>)

   at HotChocolate.Types.Interceptors.MiddlewareValidationTypeInterceptor.ValidatePipeline(ITypeSystemObject type, FieldCoordinate field, ISyntaxNode syntaxNode, IList`1 middlewareDefinitions)
   at HotChocolate.Types.Interceptors.MiddlewareValidationTypeInterceptor.OnValidateType(ITypeSystemObjectContext validationContext, DefinitionBase definition, IDictionary`2 contextData)
   at HotChocolate.Configuration.AggregateTypeInterceptor.OnValidateType(ITypeSystemObjectContext validationContext, DefinitionBase definition, IDictionary`2 contextData)
   at HotChocolate.Types.TypeSystemObjectBase`1.OnValidateType(ITypeSystemObjectContext context, DefinitionBase definition, IDictionary`2 contextData)
   at HotChocolate.Types.TypeSystemObjectBase`1.CompleteType(ITypeCompletionContext context)
   at HotChocolate.Configuration.TypeInitializer.<CompleteTypes>g__CompleteType|23_0(RegisteredType registeredType)
   at HotChocolate.Configuration.TypeInitializer.ProcessTypes(TypeDependencyKind kind, Func`2 action)
   at HotChocolate.Configuration.TypeInitializer.CompleteTypes()
   at HotChocolate.Configuration.TypeInitializer.Initialize(Func`1 schemaResolver, IReadOnlySchemaOptions options)
   at HotChocolate.SchemaBuilder.Setup.InitializeTypes(SchemaBuilder builder, IDescriptorContext context, IReadOnlyList`1 types, LazySchema lazySchema)
   at HotChocolate.SchemaBuilder.Setup.Create(SchemaBuilder builder, LazySchema lazySchema, IDescriptorContext context)
   at HotChocolate.SchemaBuilder.Create(IDescriptorContext context)
   at HotChocolate.SchemaBuilder.HotChocolate.ISchemaBuilder.Create(IDescriptorContext context)
   at HotChocolate.Execution.RequestExecutorResolver.CreateSchemaAsync(NameString schemaName, RequestExecutorSetup options, RequestExecutorOptions executorOptions, IServiceProvider serviceProvider, TypeModuleChangeMonitor typeModuleChangeMonitor, CancellationToken cancellationToken)
   at HotChocolate.Execution.RequestExecutorResolver.CreateSchemaServicesAsync(NameString schemaName, RequestExecutorSetup options, CancellationToken cancellationToken)
   at HotChocolate.Execution.RequestExecutorResolver.GetRequestExecutorNoLockAsync(NameString schemaName, CancellationToken cancellationToken)
   at HotChocolate.Execution.RequestExecutorResolver.GetRequestExecutorAsync(NameString schemaName, CancellationToken cancellationToken)
   at HotChocolate.Execution.RequestExecutorProxy.GetRequestExecutorAsync(CancellationToken cancellationToken)
   at HotChocolate.AspNetCore.HttpPostMiddleware.HandleRequestAsync(HttpContext context, AllowedContentType contentType)
   at HotChocolate.AspNetCore.HttpPostMiddleware.InvokeAsync(HttpContext context)
   at HotChocolate.AspNetCore.WebSocketSubscriptionMiddleware.InvokeAsync(HttpContext context)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

### Product

Hot Chocolate

### Version

12.1.1

**Comment by tobias-tengler** [comment]
There is no version 12.1.1 yet 🤔 I assume you are on the latest (non-preview) version 12.0.1.
Recently an issue with middleware (attribute) order in combination with ready-to-run compilation was reported: #4311 
Are you perhaps also encountering this? A fix for this was added by @michaelstaib a couple of days ago in #4337.

Could you please verify that the issue still persists with the latest preview version (12.1.0-preview.5)?

**Comment by kevpye-fabdata** [comment]
That is odd in regards to the version number as this is my nuget package manager... every hotchocolate package is at 12.1.1 - i just installed directly from the nuget package manager. I will have a look at the fix from Michael now and update after.

![hot choc](https://user-images.githubusercontent.com/82941745/140087649-f7c57b2c-d206-422f-85e4-849837d981b0.PNG)

**Comment by tobias-tengler** [comment]
Ah a version 12.1.1 was published by accident a few days ago, but it should've been unlisted, at least it is no longer on nuget.org: https://www.nuget.org/packages/HotChocolate/12.1.0-preview.5#versions-tab
Maybe the VS Package Manager still had it cashed.
Looking at your .csproj again, you are also using different versions for HotChocolate.Data and HotChocolate.Data.EntityFramework. Make sure they have the same version.

**Comment by kevpye-fabdata** [comment]
@tobias-tengler thanks so much for the crazy fast help - I had spent a few hrs crawling the docs and stack overflow trying to see if I was doing something wrong as the tutorial i'm following is using v11 and I assumed there may have been a breaking change between 11 and 12. 

The accidentally released 12.1.1 thing must have been the cause. I downgraded all of my packages to 12.0.1 and now the issue is fixed. I also refreshed my nuget package cache and 12.1.1 no longer shows up as an available option so please go ahead and close this.

Again, thanks so much for the fast replies
